### PR TITLE
Document the event auth rules

### PR DESCRIPTION
**Pull request by richvdh**
These are a necessary precursor to state resolution.

### CODE REVIEW

File: specification/server_server_api.rst [hunk 1]
```diff
@@ -442,24 +443,155 @@ keys exist to support this:
...
+         50), and the ``state_key``'s power level is less than or equal to the
+         ``sender``'s power level, then allow.
+
+      #. Otherwise, reject.
+
+   #. If ``membership`` is ``leave``:
+
+      i. If ``sender`` matches ``state_key``, allow.
+
+      #. If the ``sender``'s current membership state is not ``joined``, reject.
```
Comment (erikjohnston) [inline specification/server_server_api.rst]: There's a clause that allows an admin "unbanning" someone if they were banned an the admin has a power level above the "ban_level"

**Comment by erikjohnston** [comment]
There's also some checks to ensure that the right signatures are there: https://github.com/matrix-org/synapse/blob/master/synapse/event_auth.py#L47-L68, not sure if they should be included here or not

File: specification/server_server_api.rst [hunk 2]
```diff
@@ -442,24 +443,155 @@ keys exist to support this:
...
+
+Unauthorized events that appear in the event graph do *not* have any effect on
+the state of the room.
+
+.. Note:: This is in contrast to redacted events which can still affect the
+          state of the room. For example, a redacted ``join`` event will still
+          result in the user being considered joined.
+
+1. If type is ``m.room.create``, allow if and only if has depth 0 and it has no
+   previous events - *i.e.* it is the first event in the room.
```
Comment (erikjohnston) [inline specification/server_server_api.rst]: There's an assertion for that room_id domain matches sender domain

File: specification/server_server_api.rst (hunk 1 above)
Comment (richvdh) [reply specification/server_server_api.rst]: thanks, fixed in 47005d5.

My understanding of the code is that you can unban someone even if you have PL < ban, provided you have PL >= kick, and the target's PL is <= yours. Seems a bit odd but if it's true we might as well spec it. Can you double-check me?

File: specification/server_server_api.rst (hunk 2 above)
Comment (richvdh) [reply specification/server_server_api.rst]: yeah. I think there's a whole class of checks like this and the signatures which I want to keep separately from the auth rules, because they don't really play a part in state resolution. The line's a bit fuzzy, but I want to punt this one and the signatures to a separate section and a separate PR.

File: specification/server_server_api.rst (hunk 1 above)
Comment (erikjohnston) [reply specification/server_server_api.rst]: https://github.com/matrix-org/synapse/blob/master/synapse/event_auth.py#L320 seems to suggest you need a PL >=  ban to unban?

Comment (richvdh) [reply specification/server_server_api.rst]: Right, apparently I misread the code. Let me try again.

It looks like you have to have PL >= ban *and* PL >= kick *and* PL >= the target to unban someone, otherwise you'll get caught at https://github.com/matrix-org/synapse/blob/master/synapse/event_auth.py#L327. Right?

Comment (erikjohnston) [reply specification/server_server_api.rst]: Yup, looks like it.

Comment (richvdh) [reply specification/server_server_api.rst]: Right, one more go in 6d038e1. PTAL?

File: specification/server_server_api.rst [hunk 3]
```diff
@@ -442,24 +443,172 @@ keys exist to support this:
...
+      i. If the previous event in the room graph is an ``m.room.create``, the
+         depth is 1 and the ``state_key`` is the creator, allow.
+
+      #. If the ``sender`` does not match ``state_key``, reject.
+
+      #. If the user's current membership state is ``join``, allow.
+
+      #. If the ``join_rule`` is ``public``, allow.
+
+      #. Otherwise, reject.
```
Comment (erikjohnston) [inline specification/server_server_api.rst]: Don't we need to handle invites?

Comment (richvdh) [reply specification/server_server_api.rst]: nnnnyargh

Comment (richvdh) [reply specification/server_server_api.rst]: 2d89f1b

**Comment by erikjohnston** [comment]
<3

**Comment by richvdh** [comment]
that was harder work than it should have been. thanks!
